source/common/config/unified_mux/delta_subscription_state.cc

```diff
@@ -0,0 +1,195 @@
+      // empty resources as updates.
+      supports_heartbeats_(type_url_ != "envoy.config.route.v3.VirtualHost"), wildcard_(wildcard) {}
+
+DeltaSubscriptionState::~DeltaSubscriptionState() = default;
+
+void DeltaSubscriptionState::updateSubscriptionInterest(
+    const absl::flat_hash_set<std::string>& cur_added,
+    const absl::flat_hash_set<std::string>& cur_removed) {
+  for (const auto& a : cur_added) {
+    resource_state_[a] = ResourceState::waitingForServer();
+    // If interest in a resource is removed-then-added (all before a discovery request
+    // can be sent), we must treat it as a "new" addition: our user may have forgotten its
+    // copy of the resource after instructing us to remove it, and need to be reminded of it.
+    names_removed_.erase(a);
+    names_added_.insert(a);
+  }
+  for (const auto& r : cur_removed) {
+    resource_state_.erase(r);
+    // Ideally, when interest in a resource is added-then-removed in between requests,
+    // we would avoid putting a superfluous "unsubscribe [resource that was never subscribed]"
+    // in the request. However, the removed-then-added case *does* need to go in the request,
+    // and due to how we accomplish that, it's difficult to distinguish remove-add-remove from
+    // add-remove (because "remove-add" has to be treated as equivalent to just "add").
+    names_added_.erase(r);
+    names_removed_.insert(r);
+  }
+}
+
+// Not having sent any requests yet counts as an "update pending" since you're supposed to resend
+// the entirety of your interest at the start of a stream, even if nothing has changed.
+bool DeltaSubscriptionState::subscriptionUpdatePending() const {
+  return !names_added_.empty() || !names_removed_.empty() ||
+         !any_request_sent_yet_in_current_stream_ || dynamicContextChanged();
+}
+
+bool DeltaSubscriptionState::isHeartbeatResource(
+    const envoy::service::discovery::v3::Resource& resource) const {
+  if (!supports_heartbeats_ &&
+      !Runtime::runtimeFeatureEnabled("envoy.reloadable_features.vhds_heartbeats")) {
+    return false;
+  }
+  const auto itr = resource_state_.find(resource.name());
+  if (itr == resource_state_.end()) {
+    return false;
+  }
+
+  return !resource.has_resource() && !itr->second.isWaitingForServer() &&
+         resource.version() == itr->second.version();
+}
+
+void DeltaSubscriptionState::handleGoodResponse(
+    const envoy::service::discovery::v3::DeltaDiscoveryResponse& message) {
+  absl::flat_hash_set<std::string> names_added_removed;
+  Protobuf::RepeatedPtrField<envoy::service::discovery::v3::Resource> non_heartbeat_resources;
+  for (const auto& resource : message.resources()) {
+    if (!names_added_removed.insert(resource.name()).second) {
+      throw EnvoyException(
+          fmt::format("duplicate name {} found among added/updated resources", resource.name()));
+    }
+    if (isHeartbeatResource(resource)) {
+      continue;
+    }
+    non_heartbeat_resources.Add()->CopyFrom(resource);
```

Comment:
I have a concern with this approach (same for the `SotwSubscriptionState`). This means that for most resources the entire response will be copied before the `onConfigUpdate()` is called. This could be quite inefficient for large xDS updates. Is there a way to avoid this copy? Can't the config sinks simply ignore heartbeat resources?

Reply:
This is a close copy of what's being currently done in grpc_mux and delta_subscription_state (https://github.com/envoyproxy/envoy/blob/main/source/common/config/delta_subscription_state.cc#L105). While I see your point, I don't think we had any reports of impaired performance. 

> Can't the config sinks simply ignore heartbeat resources?

That would mean making *all* sinks aware of TTL, while in current implementation TTL implementation details are opaque to config sinks. @snowp will have more insight.

Reply:
I would go with the iterator approach. I'm worried about it, but I do not yet have perf data to back it up, as right now there other CPU burners that dominate large config updates. I suspect this may come up soon. I guess it is probably ok as it is. Can you add a TODO or comment about optimizing this copying behavior, please? Will review the rest tomorrow.

Reply:
Added todos.